File: project/jimmer-core/src/main/java/org/babyfish/jimmer/error/CodeBasedException.java
package org.babyfish.jimmer.error;

import org.jetbrains.annotations.NotNull;

import java.util.Map;

public abstract class CodeBasedException extends RuntimeException {

    protected CodeBasedException(String message, Throwable cause) {
        super(message, cause);
    }

    @NotNull
    public abstract Enum<?> getCode();

    @NotNull
    public abstract Map<String, Object> getFields();

    public ExportedError getExportedError() {
        Enum<?> code = getCode();
        return new ExportedError(
                familyName(code.getClass().getSimpleName()),
                code.name(),
                getFields()
        );
    }

    private static String familyName(String name) {
        boolean prevLower = false;
        int size = name.length();
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < size; i++) {
            char c = name.charAt(i);
            if (Character.isUpperCase(c)) {
                if (prevLower) {
                    builder.append("_");
                }
                prevLower = false;
                builder.append(c);
            } else {
                prevLower = true;
                builder.append(Character.toUpperCase(c));
            }
        }
        return builder.toString();
    }
}

File: project/jimmer-core/src/main/java/org/babyfish/jimmer/error/CodeBasedException.kt
package org.babyfish.jimmer.error

abstract class CodeBasedException protected constructor(
    message: String?,
    cause: Throwable?
) : Exception(message, cause) {

    constructor(): this(null, null)

    constructor(message: String?) : this(message, null)

    constructor(cause: Throwable?): this(null, cause)

    private val metadata: ClientExceptionMetadata =
        ClientExceptionMetadata.of(this.javaClass)

    open val fields: Map<String, Any?>
        get() = metadata
            .getterMap
            .takeIf { it.isNotEmpty() }
            ?.let {
                it.mapValues { (_, v) ->
                    v.invoke(this)
                }
            }
            ?: emptyMap()

    val family: String
        get() = metadata.family

    val code: String
        get() = metadata.code
}
